lib/uiautomator.js

```diff
@@ -1,225 +1,108 @@
+    let processIsAlive;
+    if (!startDetector) {
+      startDetector = null;
     }
-    this.onExit(code);
-  }
-};
-
-UiAutomator.prototype.checkForSocketReady = function (output) {
-  if (/Appium Socket Server Ready/.test(output)) {
-    this.socketClient = net.connect(this.systemPort, function () {
-      this.debug("Connected!");
-      this.onSocketReady(null);
-    }.bind(this));
-    this.socketClient.setEncoding('utf8');
-    var oldData = '';
-    this.socketClient.on('data', function (data) {
-      this.debug("Received command result from bootstrap");
-      try {
-        data = JSON.parse(oldData + data);
-        oldData = '';
-      } catch (e) {
-        logger.info("Stream still not complete, waiting");
-        oldData += data;
-        return;
+    try {
+      this.changeState(UiAutomator.STATE_STARTING);
+      log.info("Starting UiAutomator");
+      log.info("Parsing uiautomator jar");
+      // expecting a path like /ads/ads/foo.jar or \asd\asd\foo.jar
+      let jarName = this.parseJarNameFromPath(uiAutomatorBinaryPath);
+      await this.adb.push(uiAutomatorBinaryPath, this.tempPath);
+      // killing any uiautomator existing processes
+      await this.killUiAutomatorOnDevice();
+
+      // for bootstrapJar
+      // jarName = bootstrap.jar
+      // className = io.appium.android.bootstrap.Bootstrap
+      // extraParams = "-e pkg appPackage -e disableAndroidWatchers true/false"
+      // startDetector = (stdout) => {
+      //  if (/Appium Socket Server Ready/.test(stdout)) {
+      //       return true;
+      //  }}
+      let args = ["shell", "uiautomator", "runtest", jarName, "-c", className];
+      if (extraParams) {
+        args.push(extraParams);
       }
-      if (this.cmdCb) {
-        var next = this.cmdCb;
-        this.cmdCb = null;
-        next(data);
-      } else {
-        this.debug("Got data when we weren't expecting it, ignoring:");
-        this.debug(JSON.stringify(data));
+      const adbPath = this.adb.getAdbPath();
+      log.info(`Executing command ${adbPath} ${args}`);
+      this.proc = new SubProcess(adbPath, args);
+      // handle out-of-bound exit by simply emitting a stopped state
+      this.proc.on('exit', (code, signal) => {
+        processIsAlive = false;
+        // cleanup
+        if (this.state !== UiAutomator.STATE_STOPPED &&
+            this.state !== UiAutomator.STATE_STOPPING) {
+          let msg = `UiAutomator exited unexpectedly with code ${code}, ` +
+                    `signal ${signal}`;
+          log.error(msg);
+        }
+        else if (this.state === UiAutomator.STATE_STOPPING) {
+          log.debug("UiAutomator shut down normally");
+        }
+        this.changeState(UiAutomator.STATE_STOPPED);
+      });
+      await this.proc.start(startDetector);
+      processIsAlive = true;
+      this.changeState(UiAutomator.STATE_ONLINE);
+      return this.proc;
+    } catch (e) {
+      this.emit(UiAutomator.EVENT_ERROR, e);
+      if (!processIsAlive) {
```

Comment:
this logic seems backward to me. don't we only want to kill if the processIsAlive?